Make sure all UI tests have at least 1x verifyHtml and the rest can be verifyMainHtml #3709

MrFool opened the pull request:
Fixes #3709

Almost done, will ping when it's done, still need to regenerate for those who used to test too much or too little.

MrFool commented:
Need some help with reviewing this, now all the pages that have had their tests implemented already (note some admin pages don't even have tests) have at least 1x verifyHtml and the rest verifyHtmlMainContent, so the reason why this has so much changes is because some stuff that used to not have verifyHtml now have verifyHtml and that adds a lot of changes.

wkurniawan07 commented on src/test/resources/pages/unregisteredStudentFeedbackResultsPageOpen.html:
```diff
@@ -274,7 +274,7 @@
                                                                 <!--Note: When an element has class text-preserve-space, do not insert and HTML spaces-->
-                                                                <td class="text-preserve-space">Team 2 (charlie) to team 1</td>
+                                                                <td class="text-preserve-space">Team 2 (danny) to team 1</td>
                                                             </tr>
```
Whoa. This got you. Be careful. Or... is this intended?

MrFool replied:
I have no idea, I most likely missed this out when looking through, this is the Danny and Charlie not getting along thing yeah? They like to switch places from time to time, perhaps I can revert this file?

MrFool commented:
I reverted that file, do continue when there is time :)

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java:
```diff
@@ -54,7 +54,9 @@ public void testEditResponse() {
-        submitPage.verifyHtml("/InstructorEditStudentFeedbackPageOpen.html");
+
+        // Full HTML verification already done in InstructorFeedbackSubmitPageUiTest
+        submitPage.verifyHtmlMainContent("/InstructorEditStudentFeedbackPageOpen.html");
```
This is changed, but the HTML file is untouched. God mode doesn't touch the file if the parts to be compared is the same (even if the HTML file is full file, if we pass to `verifyHtmlMainContent`, God mode will only compare the `mainContent` part).
If you want God mode to regenerate it, mess up with the `mainContent` part (can be as simple as adding random characters) then regenerate.

MrFool replied:
Ahh, I see, ok I'll get to that.

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java:
```diff
@@ -82,8 +84,9 @@ private void testAddResponse() {
-        submitPage.verifyHtml("/InstructorEditStudentFeedbackPageModified.html");
+
+        // Full HTML verification already done in InstructorFeedbackSubmitPageUiTest
+        submitPage.verifyHtmlMainContent("/InstructorEditStudentFeedbackPageModified.html");
```
And here too.

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/StudentCourseJoinConfirmationPageUiTest.java:
```diff
@@ -175,7 +177,7 @@ private void testJoinConfirmation() throws Exception {
         confirmationPage = createNewPage(browser, StudentCourseJoinConfirmationPage.class);
         // this test uses accounts from test.properties
-        confirmationPage.verifyHtml("/studentCourseJoinConfirmationHTML.html");
+        confirmationPage.verifyHtmlMainContent("/studentCourseJoinConfirmationHTML.html");
```
And here too.

wkurniawan07 commented:
Some comments:
1. I think you missed instructorCourseEnrollResult.jsp.
2. On the instructorFeedbackResultsPageUiTest, strictly speaking, you should do full verification for _every single_ view type possible because they're rendered with different jsp files. But I kind of understand if you don't want to do this, I think this is unnecessary repetition also.
3. Are there reasons for distinguishing registered and unregistered student in some of the pages?

damithc commented on src/test/resources/pages/InstructorCourseEnrollPage.html:
```diff
@@ -1,34 +1,293 @@
-<div class="container" id="mainContent">
```
Some of these changes to html files do not correspond to the purpose of the PR.

MrFool replied:
You mean the br and the slashes and those?

MrFool replied:
They are actually added because I used god mode to regenerate them, god mode can be quite erratic sometimes and add in odd indentations or slashes etc, should I manually revert those?

damithc commented:
A general note: I don't mind having 'more than enough' full html verifications. Reasons: 
1. partial verification weakens the test. Sometimes we expected one part of the the page to change and test for that part only, but a bug in the code can change a different part of the page 'unexpectedly'. Full html verification catch those bugs too. 
2. Because we are now using GodMode to generate html pages, modifying html files is not as difficult as before, so the cost of doing a full verification is not as high as before.
3. partial verification uses regex, which can slow down testing.
4. Future maintenance can remove the full verification for valid reasons (e.g. because that test is no longer relevant).

MrFool commented:
@damithc 

Actually, this PR increases the number of full HTML verifcations :p After doing the PR I realised that there were pages that doesn't have a single full HTML verification so I added them more than I removed them.

As for removing accidentally, I added comments for the full HTML verifications to signify that they are the full HTML verification, if they should be removed (at the very least), another full verification should be made. I'm quite certain that full verifications have been removed accidentally before, which is why some pages didn't have full HTML verification when I was working on this issue. The comments should help out in this department.

As for whether or not to change to more full HTML verification or whether we should investigate the cost between partial checks and full checks, you can let us know and we can look at it. But currently it seems that the consistent thing to do across all ui tests is to have 1x full verification and the rest partial verification. I could do it if you want - to investigate the costs of partial versus full verification and which we should use more of.

@wkurniawan07 

I missed that out indeed, it was hidden inside the enroll page, thanks for pointing that out!

They are currently rendered with different jsp files but tags will solve that issue I hope? Or is that not the case, if it is not I can look into doing that.

Yes there is a reason for doing so, previously there was a discrepancy between what would be seen by registered versus non registered users (the footer) and was fixed by yoh jun, we want to distinguish between the two of them so we can make sure regression doesn't happen.

damithc commented:
Yes, we must have at least 1 full html verification. so this PR is in the right direction. I'm saying 1 should be a lower limit only. More is not necessarily bad. When in doubt, do a full html verification.

MrFool commented:
I understand, I just needed to explain what the PR was doing since +4.7k lines and -2k lines might not really explain what is going on :s

I did see a comment in one of the pages once, saying:

"Make sure at least one verifyHtml, make the rest verifyHtmlMainContent"

should we have a better guideline for this because without it future maintainers might not actually know about it and remove it unintentionally and it could slip through reviews?

MrFool commented:
Manually changed the files that I missed and added instructor course enroll and checked for dev green, do continue reviewing when you're free.

On a side note, there are a lot of "unrelated changes" because god mode seems to make a lot more changes than just the actual changes, esp when it comes to indentation and break lines.

wkurniawan07 commented:
Looks good. Can be merged as long as dev green.
